pkg/services/promtypemigration/prom_mig.go

```diff
@@ -0,0 +1,82 @@
+package promtypemigration
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+
+	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/plugins/manager/registry"
+	"github.com/grafana/grafana/pkg/plugins/repo"
+	"github.com/grafana/grafana/pkg/services/datasources"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+type PromMigrationHandler interface {
+	Migrate(context.Context, *promMigrationService) error
+}
+
+type promMigrationService struct {
+	cfg                *setting.Cfg
+	dataSourcesService datasources.DataSourceService
+	pluginRegistry     registry.Service
+	pluginRepo         repo.Service
+	pluginInstaller    plugins.Installer
+}
+
+func (s *promMigrationService) applyMigration(ctx context.Context, pluginID string, promDataSources []*datasources.DataSource) error {
+	if len(promDataSources) == 0 {
+		return nil
+	}
+
+	// check to see if prom is installed, if not install it
+	if _, installed := s.pluginRegistry.Plugin(ctx, pluginID, ""); !installed {
+		compatOpts := plugins.NewAddOpts(s.cfg.BuildVersion, runtime.GOOS, runtime.GOARCH, "")
+		err := s.pluginInstaller.Add(ctx, pluginID, "", compatOpts)
+		if err != nil {
+			return err
+		}
```

Comment:
you may not want to return an error here. That will break, for example, offline scenarios in which people don't have access to the catalog. My suggestion is to log the error instead and omit the migration. Then in Grafana 13 or the next major version, when you are ready to make a breaking change you could return an error here if you want to.

Reply:
I think we want it to break here. If it gets here its because we don't have the data source and we dont want to continue with the migration if the grafana doesnt have the data source